### PR TITLE
Add templated bitwise AND block with tests

### PR DESCRIPTION
**Pull request by alexhojinpark**
Refs #161

Ported the bitwise AND functionality from GNU Radio 3.10 to GNU Radio 4 as a templated `And` block, enabling support for multiple integer types via templates. 

The implementation is covered by Boost.UT unit tests, validating basic operations, edge cases, and boundary conditions.

Changes:
- Added: `blocks/basic/include/gnuradio-4.0/basic/And.hpp`, `blocks/basic/test/qa_And.cpp`
- Updated: `blocks/basic/CMakeLists.txt`, `blocks/basic/test/CMakeLists.txt`

The current tests cover key scenarios. Should I extend them to include additional bit patterns?

Updated (2025-04-18):
- Added More Blocks: `AndConst.hpp`, `Not.hpp`, `Or.hpp`, and `Xor.hpp`
- Added Test for Blocks: `qa_AndConst.cpp`, `qa_Not.cpp`, `qa_Or.cpp`, and `qa_Xor.cpp`
- Updated: `blocks/basic/CMakeLists.txt`, `blocks/basic/test/CMakeLists.txt`

### CODE REVIEW

File: blocks/basic/include/gnuradio-4.0/basic/And.hpp [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+#include <gnuradio-4.0/Block.hpp>
+#include <gnuradio-4.0/BlockRegistry.hpp>
+
+namespace gr::basic {
+
+GR_REGISTER_BLOCK(gr::basic::And, [uint8_t, int16_t, int32_t])
+
+template<typename T>
+requires std::is_arithmetic_v<T>
```
Comment (RalphSteinhagen) [inline blocks/basic/include/gnuradio-4.0/basic/And.hpp:12]: `std::is_arithmetic_v<T>` also includes. This is probably not what you intended. 

Maybe the concept [std::integral](https://en.cppreference.com/w/cpp/concepts/integral) might be a better choice, i.e. `template<std::integral>`.

File: blocks/basic/include/gnuradio-4.0/basic/And.hpp [hunk 2]
```diff
@@ -0,0 +1,29 @@
+#ifndef AND_HPP
+#define AND_HPP
+
+#include <gnuradio-4.0/Block.hpp>
+#include <gnuradio-4.0/BlockRegistry.hpp>
+
+namespace gr::basic {
+
+GR_REGISTER_BLOCK(gr::basic::And, [uint8_t, int16_t, int32_t])
```
Comment (RalphSteinhagen) [inline blocks/basic/include/gnuradio-4.0/basic/And.hpp:9]: OK, maybe it might be worthwhile to expand this compared to the [GR3 version](https://wiki.gnuradio.org/index.php?title=And) to cover all signed and unsigned integral types? 

@mormj what do you think? #GSoC

File: blocks/basic/include/gnuradio-4.0/basic/And.hpp [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    GR_MAKE_REFLECTABLE(And, in1, in2, out);
+
+    [[nodiscard]] constexpr T processOne(T input1, T input2) const noexcept {
+        return input1 & input2;
+    }
+};
+
+} // namespace gr::blocks
+
+#endif // AND_HPP
```
Comment (RalphSteinhagen) [inline blocks/basic/include/gnuradio-4.0/basic/And.hpp:29]: minor: an empty line at the end of the file is missing.

File: blocks/basic/test/qa_And.cpp [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+            & 0000 0000 0000 0011   (3 in binary)
+            --------------------
+            0000 0000 0000 0011   (Result = 3)
+         */
+        expect(eq(andBlock.processOne(static_cast<int16_t>(-5), static_cast<int16_t>(3)), static_cast<int16_t>(3)));  // -5 & 3 = 3
+
+    };
+};
+
+int main() { return boost::ut::cfg<boost::ut::override>.run(); }
```
Comment (RalphSteinhagen) [inline blocks/basic/test/qa_And.cpp:61]: minor: empty line at eof missing.

**Review by RalphSteinhagen — CHANGES_REQUESTED** [review]
@alexhojinpark thanks for your interest in and first contribution to GR4! :+1: 

The code already looks good as is and seems to be well tested. There are some minor comments (see above). 

If you'd be willing, would you implement the full set of `'Boolean Operators' (5 blocks)` outlined in the [EPIC#161](https://github.com/fair-acc/gnuradio4/issues/161)? 

i.e. 
  * [And Const](https://wiki.gnuradio.org/index.php?title=And_Const)
  * [Not](https://wiki.gnuradio.org/index.php?title=Not)
  * [Or](https://wiki.gnuradio.org/index.php?title=Or)
  * [Xor](https://wiki.gnuradio.org/index.php?title=Xor)
  
This would go a long way! I like the approach of templating like done, for example, in [here](https://github.com/fair-acc/gnuradio4/blob/983d49314931d624c363b4be176b5b93492e27c8/blocks/math/include/gnuradio-4.0/math/Math.hpp#L29-L56) but this is not a must.

In any case, thanks for your PR and let me know if you are OK with implementing the other similar functions as well. Much appreciated.

Otherwise, I'd merge this as is (+ the small changes outlined above).

**Comment by alexhojinpark** [comment]
@RalphSteinhagen Thank you for the feedback! Yes, I’d be happy to continue working on implementing the full set of 'Boolean Operators' for EPIC#161.

Let me know if there are any specific considerations I should keep in mind. Looking forward to contributing more!

**Comment by RalphSteinhagen** [comment]
@alexhojinpark could you perhaps rebase and (force-)push (possibly with your other additions) again? I changed the restyled bot to (hopefully) accept external public PRs. The previous rule was a bit too restrictive. To become active, you'd need to rebase to the present head in 'main' though. Thanks in advance.

**Comment by codblack789** [comment]
@alexhojinpark how did u build the gnuradio4 locally.. and set it up, can u explain ?

**Comment by alexhojinpark** [comment]
@codblack789 Hi, For setting up GNURadio4, I’d check the [Building section](https://github.com/fair-acc/gnuradio4?tab=readme-ov-file#building) in the README. It walks you through dependencies and CMake setup.

Once you’ve got the build dir ready (`gnuradio4/build/`), you can run `make` there to build it locally. Hope it answered your question!

**Comment by codblack789** [comment]
@alexhojinpark 
Yeah I followed the same steps 
1. Clone 
2. cd gnuradio4 
3.mkdir build 
4.cd build

5.cmake -DCMAKE_BUILD_TYPE=RelWithAssert -DENABLE_BLOCK_PLUGINS=ON -DENABLE_BLOCK_REGISTRY=ON .. 
6  --build . -- -j$(nproc)

After that when u open the same in VS code, I get Errors near header files !!

No idea where am I going wrong !!

**Comment by RalphSteinhagen** [comment]
> @alexhojinpark Yeah I followed the same steps
> 
> 1. Clone
> 2. cd gnuradio4
>    3.mkdir build
>    4.cd build
> 
> 5.cmake -DCMAKE_BUILD_TYPE=RelWithAssert -DENABLE_BLOCK_PLUGINS=ON -DENABLE_BLOCK_REGISTRY=ON .. 6 --build . -- -j$(nproc)
> 
> After that when i open the same in VS code, I get Errors near header files !!
> 
> No idea where am I going wrong !!

**Comment by KrxGu** [comment]
@alexhojinpark 

Hey there! I noticed the CI checks are failing for formatting (Restyled) and Sonar. Since I don’t have permission to push changes to your branch(@RalphSteinhagen :( ), I wanted to offer some pointers:

**Restyled (clang-format)**: You can apply the suggested formatting locally by running something like:

```
clang-format -i blocks/basic/include/gnuradio-4.0/basic/And.hpp blocks/basic/test/qa_And.cpp
git commit -am "Apply clang-format suggestions"
git push
```

This should resolve the restyled check.

**Sonar**: Sometimes forks can’t access Sonar credentials, so it may be normal for that check to fail from an external PR. You might ask the maintainers if it can be ignored or if they have a workaround.

Let me know if you need more details—happy to help walk you through the steps!

**Comment by RalphSteinhagen** [comment]
@KrxGu and @alexhojinpark alternatively you can check the GitHub restyled [action output](https://github.com/fair-acc/gnuradio4/actions/runs/14123484603/job/39569504719?pr=536#step:5:149) that indicates what and how to apply missing restyled diff.

Small edits aside, we cannot fix remote branches (N.B. @alexhojinpark is pushing from his forked repo). That's normal ...

File: blocks/basic/include/gnuradio-4.0/basic/And.hpp (hunk 2 above)
Comment (mormj) [reply blocks/basic/include/gnuradio-4.0/basic/And.hpp:10]: Yes, I would expand this out to at least the types supported in GR3

**Comment by alexhojinpark** [comment]
@KrxGu @RalphSteinhagen 

Thank you for the suggestions. I'll make sure to follow the restyled format, and push with all other `Boolean Operators (5 blocks)` which I am working on right now. :)

**Comment by RalphSteinhagen** [comment]
@alexhojinpark do you plan to continue working on this PR? If yes, you may need to rebase against the latest main commit.

**Comment by alexhojinpark** [comment]
@RalphSteinhagen Yes, I’ve rebased and force-pushed accordingly!

**Comment by mormj** [comment]
Hi @alexhojinpark - really appreciate the effort in working through the process of creating a new GR4 block.  I would recommend closing this in favor of #595 as it has covered a larger number of similar blocks and pushed through a number of CI issues.
